### PR TITLE
feat: add dkoverlay, ltoverlay and nooverlay attributes

### PR DESCRIPTION
**Pull request by diana-wang**
# Alaska Airlines Pull Request

**Fixes:** #12 
## Summary:

Updates dialog overlay opacity to support dkoverlay, ltoverlay, nooverlay settings. Default for default dialog state is ltoverlay. Default for modal/blocking state is dkoverlay.

## Type of change:

- [x] New capability
- [x] Revision of an existing capability

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

**Comment by blackfalcon** [comment]
I have many reservations about this. The intentional use of the light and dark backgrounds for the dialog was to be symbolic as to the type of dialog box being used. E.g. light is for dialog and dark is for the modal dialog. 

This update breaks that pattern very intentionally and leaves the use of these backgrounds open for other-than intended interpretation and inconsistent use across the various platforms. 

What is the intent behind this, what is the design direction? 

@Crea-Ive  and @vidalmenAS I feel you need to weigh in on this as well.

File: src/style.scss [hunk 1]
```diff
@@ -50,6 +50,30 @@ $auro-spacing-options: none;
   }
 }
 
+:host([dkOverlay]) {
```
Comment (blackfalcon) [inline src/style.scss:53]: I can't agree with this. The selectors written here dismiss the intent of the options.

**Comment by diana-wang** [comment]
#12 describes the reasoning behind this PR, but it sounds like there needs to be additional discussion between the designers. I'll let @clockeby continue the conversation there.

If a different way of adding this feature is agreed on from that discussion, I'd be happy to implement it differently.

File: src/style.scss (hunk 1 above)
Comment (diana-wang) [reply src/style.scss:53]: Noted. I'll modify these changes as needed based on what comes out of the discussion.

**Comment by clockeby** [comment]
The intent is to add more flexibility for other use cases. I talked to Vidal about it initially. He said his original recommendation was to use a simple drop shadow but due to the scenario needed for shopping (flight results page) the background was too busy and a drop shadow alone didn't allow for enough focus for the modal content, which I agree with given the application of the modal in that case. For the blocking modal, darkening the background also makes sense. It's a full stop until you make an A or B decision.

For the seat map, we'd like to use it for displaying seat-specific info (upgrade price, no window, limited recline, etc.) and for those instances, even the lighter, non-blocking background feels heavy. It darkens the entire page (which is split to allow for focused functionality in two panes) to let you know a minimum of information. We can certainly use it, but the original, more minimal recommendation seems much more fitting.

**Comment by blackfalcon** [comment]
Terms clarity. The component is auro-dialog. All instances of this component are dialog boxes. In the case that the user is forced to interact with the dialog, this is referred to as the modal dialog. 

What you are describing here actually is starting to sound less and less like a dialog and more like a popover. The purpose of a dialog box is to bring attention to the content being presented. E.g. a dismissible alert or other important information that the user must know. 

This feels like a case where the additional content is supplemental to their decision making. E.g. another option with popover or an accordion. 

In the past, there has been a tendency to move all non-page content into dialog boxes. Maybe this is an opportunity to revisit that?

**Comment by diana-wang** [comment]
We considered using auro-popover, but decided against it since its primary scope was for tooltips and helper text shown on hover on desktop. Whereas the ones we need for seatmaps require them to be shown on click and to be supported on mobile. Based on what you said, this does seem like a scenario that's between the two components.

**Comment by clockeby** [comment]
From what I've read, dialogs are essentially non-blocking in nature. That is, the background is not disabled and they can be interacted with, ignored, moved, etc. Supplemental information but not essential enough to _require_ interaction. A modal-dialog increases the importance requiring interaction or confirmation in order to proceed. Therefore the background is disabled and often darkened to focus attention and prevent forward movement.
[https://www.nngroup.com/articles/modal-nonmodal-dialog/]

Feels like a which came first situation. Had the initial dialog work been done by myself and the seats team it would be using Vidal's original recommendation of a simple drop shadow and we'd be having a conversation about adding a darkening scrim so shopping could use it more effectively. Either way, similar needs, just not exactly alike.

**Comment by blackfalcon** [comment]
WARNING ... this is a pretty fluid thought, sorry. 

> We considered using auro-popover, but decided against it since its primary scope was for tooltips and helper text shown on hover on desktop. Whereas the ones we need for seatmaps require them to be shown on click and to be supported on mobile. Based on what you said, this does seem like a scenario that's between the two components.

That is an implementation detail. The concept of a popover is to have a small interaction of additional content triggered by the main content. The trigger doesn't 'have' to be a rollover, it can be a click. 

> Feels like a which came first situation. Had the initial dialog work been done by myself and the seats team it would be using Vidal's original recommendation of a simple drop shadow and we'd be having a conversation about adding a darkening scrim so shopping could use it more effectively. Either way, similar needs, just not exactly alike.

I disagree. Virtually every UI system that I am are of, iOS and otherwise, the use of a dialog/modal is an interaction model that requires the user's interaction one way or another. This, by design, has an element that separates the dialog from the content to draw attention to the content. What's interesting is that the use case for popover with iOS even does this. What's also interesting is that the popover is to be never used for handheld experiences where that should be moved to a modal dialog. 

My point is that a dialog should really only be used for use cases where the user's attention is required. Now, if that attention demands interaction or not, is the option. 

A popover on the other hand is designed to be passive. Additional controls. Supplemental content. The WC auro-popover was first developed as a 'tooltip' but that's not its limitation. It's almost sounding like to me that the popover could have different types, `tooltip` being one of them. 

Looking at the structure of the popover, it can accept any content inside the `<slot>` element. Also, the current implementation supports the trigger as a rollover. This can easily be updated to support a click trigger. Again, this click, as shown in the examples of the Auro site, can be bound to anything. Button, anchor, icon, etc ... 

Screen Shot 2020-10-07 at 5 23 46 PM

When you really break this down, the difference between a popover and a dialog is so subtle that I think that it's the intention of the user experience that we need to really define and let the design help enforce that. 

E.g. dialogs and modal dialogs; the intention is to interrupt the user from the current contextual flow. Alert. In important information to consider. Demanding a choice being made before moving forward. 

Popovers on the other hand are more passive. Never used for alerts. Always triggered by interaction. Never demanding a choice being made, but options to offer are acceptable. 

This conversation makes me think about larger context issues that were not considered in the building of the auro-popover. Being, how does this interact with mobile? The use case as driving for the 1.0 version was purely desktop. 

I would suggest that popovers on mobile will only appear on a click. Right now this defaults to a hover, but you can click on an element on mobile and this will trigger the hover event. This really isn't a good or stable experience. Also, if a popover is triggered on a hover, do you want this to even appear on mobile? Last, if it appears on mobile, how should it? Revisiting the iOS guidelines, I would argue that all popovers should present themselves as a medium dialog on mobile as it's max. 

#### My recommendations for the conversation 

1. The auro-dialog stays as is. The weight of the background is directly tied to the intent of the dialog
1. The auro-popover is expanded to support a click event as well as hover
1. auro-popover needs to have an option to hide from mobile 
1. auro-popover needs to present itself as a lower drawer on mobile 

These suggestions are not meant to address your issue @clockeby, but really make popover better. 

I 100% agree that this is a close as it gets. What I am striving for here is clarity of use. What I would like to see is a series of questions about the interaction and there be clear answers as to what component to use. 

> Does the content in the interaction require attention?
> If the content is never seen, does it matter?
> Is the content supplemental? 

Looking one last time at iOS, what is also interesting is that all interactions that require a user's attention away from the presented flow of content is a modal dialog. All these interactions, no matter how small, are presented with a heavy backdrop. 

Walking through all of this I have to say, that given what I know about your use case, this is in fact a dialog. And by all rights, that should have a backdrop. As the intent is to pull the user's intention away from the flow of content and focus on the task presented.

**Comment by diana-wang** [comment]
Per a conversation with @vidalmenAS @blackfalcon and @clockeby, we'll be moving forward with @blackfalcon's suggestion above (leaving auro-dialog as it is and expanding auro-popover). Due to that, I will be closing this PR.

**Pull request closed without merging**